Use copy_file_range(2) to implement System.Native.CopyFile()

lpereira opened the pull request:
From the man page:

       The  copy_file_range()  system  call performs an in-kernel
       copy between two file descriptors without the addi‐ tional
       cost of transferring data from the kernel to user space and
       then back into the kernel.

       (...)

       copy_file_range()  gives  filesystems  an opportunity to
       implement "copy acceleration" techniques, such as the use of
       reflinks (i.e., two or more inodes that share pointers to the
       same copy-on-write disk blocks) or server-side-copy (in the
       case of NFS).

This builds on the work by @filipnavara to copy files using `clonefile()` on macOS and on my previous PR in the [corefx repo](https://github.com/dotnet/corefx/pull/39235).

lpereira committed:
System.Native: Break SystemNative_CopyFile() into smaller functions

This is in preparation to introduce a different method of copying files
on Linux, using the copy_file_range(2) system call.

lpereira committed:
System.Native: Use copy_file_range(2) on Linux to copy files

From the man page:

       The  copy_file_range()  system  call performs an in-kernel
       copy between two file descriptors without the addi‐ tional
       cost of transferring data from the kernel to user space and
       then back into the kernel.

       (...)

       copy_file_range()  gives  filesystems  an opportunity to
       implement "copy acceleration" techniques, such as the use of
       reflinks (i.e., two or more inodes that share pointers to the
       same copy-on-write disk blocks) or server-side-copy (in the
       case of NFS).

lpereira committed:
System.Native: Cache availability of sendfile() system call

lpereira committed:
System.Native: Refactor file metadata copying when copying files

lpereira committed:
System.Native: Use SystemNative_Open() to handle O_CLOEXEC et al

There's already code to handle the existence of O_CLOEXEC in the
system, among other things, so just call the function to clean things
up slightly.

stephentoub commented:
> This builds on the work by @filipnavara to copy files using clonefile() on macOS and on my previous PR in the corefx repo.

Unfortunately that introduced a bug we still need to sort out:
https://github.com/dotnet/corefx/issues/42455
I'm less than happy with the current state of things.

lpereira commented:
> Unfortunately that introduced a bug we still need to sort out:
> [dotnet/corefx#42455](https://github.com/dotnet/corefx/issues/42455)

Let me see if I can reproduce this issue here.

lpereira commented:
@stephentoub @filipnavara Now that `copyfile()` is in place and it doesn't take file descriptors -- and only file paths --, is there any value in using a `FileStream` in `System.IO.FileSystem.CopyFile()`, other than providing a convenient way to lock the source file while it's being copied?  (Not even considering the fact that the abstraction is now broken and things are doing more work than necessary on macOS.)

filipnavara commented:
> is there any value in using a `FileStream` in `System.IO.FileSystem.CopyFile()`, other than providing a convenient way to lock the source file while it's being copied? (Not even considering the fact that the abstraction is now broken and things are doing more work than necessary on macOS.)

It does the error and POSIX advisory lock handling for the source file. There's no easy way around it on macOS without breaking the existing behavior.

lpereira committed:
System.Native: Perform locking within CopyFile()

Since all the different copying methods are different across supported
operating systems, it's better to move this logic to where the rubber
meets the road.

This replicates the behavior of the FileStream instance that used to be
used by the managed portion of this API.

lpereira commented:
> > is there any value in using a `FileStream` in `System.IO.FileSystem.CopyFile()`, other than providing a convenient way to lock the source file while it's being copied? (Not even considering the fact that the abstraction is now broken and things are doing more work than necessary on macOS.)
> 
> It does the error and POSIX advisory lock handling for the source file. There's no easy way around it on macOS without breaking the existing behavior.

Looked at this last week, and it's not that complicated to perform the locking in the C side of things, provided that, if `open()` fails, the managed portion knows that it was `open()` that failed, so it throws the correct exception (dir-not-found vs file-not-found).  I have some patches here that clean up the file copying quite a bit and implements this (by using negative `errno` for the `open()` case, positive `errno` for the file copying error, and `0` for success), and tests are passing on Linux.   Will try it on a Mac this week and if it's working fine, will update this PR.

filipnavara commented:
@lpereira Thanks for looking into it. I had a patch that moved the locks to native code (https://github.com/dotnet/corefx/issues/42455) but it was not nice to duplicate the logic between managed code and the native code. I wanted to make some tests before proceeding further but due to my temporary relocation I am bit short on time and access to my test machines is severely limited.

lpereira commented:
> @lpereira Thanks for looking into it. I had a patch that moved the locks to native code ([dotnet/corefx#42455](https://github.com/dotnet/corefx/issues/42455)) but it was not nice to duplicate the logic between managed code and the native code. I wanted to make some tests before proceeding further but due to my temporary relocation I am bit short on time and access to my test machines is severely limited.

I force-pushed a new version that moves locking to native code but keeps changes to the managed code to a minimum.  It also cleans up a few more things.  `System.IO.FileSystem.Tests` (in macOS and Linux) are passing.  I'm probably going to add the test from [dotnet/corefx#42455](https://github.com/dotnet/corefx/issues/42455) to this suite as well.

lpereira commented:
/azp run

lpereira commented:
/azp run runtime

lpereira commented:
/azp run runtime-libraries

filipnavara commented:
I do like how the code is restructured. Unfortunately it seems to be failing on two of the unit tests with an assert in native code:
```
Assertion failed: (false && "assert_err failed"), function SystemNative_ConvertErrorPalToPlatform, file /Users/runner/runners/2.164.8/work/1/s/src/libraries/Native/Unix/System.Native/pal_errno.c, line 397.
```

filipnavara commented on src/libraries/Native/Unix/System.Native/pal_io.c:
```diff
@@ -1362,74 +1378,74 @@ static bool CopyFile_CopyFileRange(int inFd, int outFd, const struct stat_ *sour
-    struct stat_ destStat;
-    while ((ret = stat_(destPath, &destStat)) < 0 && errno == EINTR);
-    if (ret == 0)
+    if (SystemNative_FLock(inFd, LOCK_EX | LOCK_NB) < 0)
```
Should not this be unlocked somewhere later?

lpereira replied:
No. See comment down below: when the file is closed, the lock is dropped.

filipnavara replied:
Fair point. The man agrees with that although I remember having to do that manually on macOS when I was testing my original PR 🤷‍♂️. I was trying to figure out why some of the unit tests are still failing.

lpereira replied:
I wouldn't be surprised, though, if macOS had a, uh, unique behavior here.  Wouldn't be the first time I encounter such things.

`flock()` operates on file descriptions rather than file descriptors, and since the description here is locked exclusively (as opposed to a shared lock), no other thread should be able to obtain that lock.  So even if another thread successfully opens the file, the `flock()` call should fail (and closing the file descriptor when this call fails will only drop the file description reference once, leaving the ref for the locked descriptor to close).  None of the tests are duplicating file descriptions, so even if another thread opens the source file again with `open()` this should be an issue.  However, if `flock()` on macOS behaves differently, treating a second `open()` in a file path as if it were a `dup()` on an already open/locked file, then things will be hairy to fix...

However,  if I'm reading this right, these failures seem to be related with not being able to upload results to Helix.  Let me try rebasing and pushing again to trigger a rebuild.  Let's see what happens from there.

filipnavara replied:
> However, if I'm reading this right, these failures seem to be related with not being able to upload results to Helix. Let me try rebasing and pushing again to trigger a rebuild. Let's see what happens from there.

Nah, the test failures were relevant:
https://dev.azure.com/dnceng/public/_build/results?buildId=519601&view=ms.vss-test-web.build-test-results-tab

filipnavara replied:
All the test failures are some form of `The process cannot access the file 'xxx' because it is being used by another process.`

lpereira replied:
Just force-pushed a new revision that explicitly calls `flock(LOCK_UN)` before closing the file. Let's see what the bots will tell us.

lpereira replied:
Runtime-live-build seems to have passed on macOS. :)

filipnavara replied:
The errors are coming only now :-/ It may take a while until all of them crop up with stack traces on Azure Pipelines test dashboard.

lpereira replied:
Yeah. They're probably going to be the same as before. Will try to carve some time for this PR sometime this or the next week.

filipnavara replied:
I'll try to reproduce it tomorrow on my macOS machine and report back.

filipnavara commented:
On a related note, it still suffers from the lack of lock check on the destination file (like I tried to bring back in https://github.com/dotnet/corefx/pull/42458). That is not unit tested at the moment and hence it is unrelated to the CI failures.

filipnavara commented:
I locally fixed the issues mentioned above and submitted a PR to test it on the CI. Please feel free to cherry-pick the changes from the linked PR if it succeeds.

carlossanlop commented:
@lpereira @filipnavara what is the status of this PR and the [draft submitted by Filip](https://github.com/dotnet/runtime/pull/33159)? Can the draft be closed?

@lpereira will you still be able to address this change?

carlossanlop commented:
@lpereira moved to another team so I'm not sure he will have a chance to continue addressing this change. Let's go ahead and close this PR. @filipnavara let's make your PR the main one to fix this issue.
